### PR TITLE
WIP: Refactor IntSets to use BitVectors

### LINKED ISSUE #3039: IntSet should be a wrapper around a BitArray
This will, in particular, allow the primes function to construct the sequence of integers as a BitArray and then return an IntSet, which is a more desirable API.

### PR DESCRIPTION
**Pull request by mbauman**
This WIP is to merge my [IntSets.jl](https://github.com/mbauman/IntSets.jl) project into base.  Here's what this entails:
- Use BitVectors instead of a mismash of bitvector.c and ad-hoc Julia code.
- IntSet is much better defined to only hold integers in `0:typemax(Int)-1`.  Before it wasn't consistently defined, particularly when it came to complement IntSets.
- Invert the complement IntSet semantics.  Instead of filling ones outside the range of the IntSet, simply reverse what a set bit means.  This means that inverting an IntSet is O(1) instead of O(n) and fixes a few bugs.
- For mathematical functions (intersect, union, etc.), use BitVector functions to work 64 bits at a time instead of working bit-by-bit.
- Increase test-coverage to 100%
- Would close #3039 (with a new constructor)

~~There are still two outstanding issues here that I could use some help with:~~ The only thing left is a performance regression:
- [x] ~~Julia fails to launch with a segfault when the sysimg.{dll/dylib/so} is unavailable.  It happens during the compilation of an [inference function that makes heavy use of IntSets](https://github.com/JuliaLang/julia/blob/bf58fa6b39844e1e93f501e100aaf94a9bc27dbe/base/inference.jl#L1294-L1321), but I can't tell _why_ with my rudimentary LLDB/LLVM knowledge.  My hunch is that I've missed some functions that are needed but load after inference.jl in the sysimg, but it successfully bootstraps… so I'm not sure.~~ The new inference bootstrap separation made this a cinch to debug.  The culprits were some redundant `Base.` module-qualified names (since Base doesn't exist yet).
- [ ] While many functions are way faster, there's still a performance gap for a few functions; most notably ~2x for `in`.  See these albums for a comparison against the current master with varying [density](http://imgur.com/a/XWrS1) and [size](http://imgur.com/a/guwui).  Some of this is due to the addition of an unnecessary GC frame (#9974), but I think most of it is from the extra dereference and perhaps some more-frequent cache misses.  Particularly confusing to me is `in` — there are fewer branches and the LLVM/native code looks much better (especially once I manually inline the getindex call to eliminate the GC frame), but it still lags in performance.  That said, it doesn't seem to have an effect on overall compilation times from looking at the linalg test suite.

### CODE REVIEW

**Comment by JeffBezanson** [comment]
Very nice. The thorough performance comparison is pretty impressive!

**Comment by tkelman** [comment]
Remember to add `intset` to `test/runtests.jl`

**Comment by mbauman** [comment]
Interesting thought from #1032: **If** we can somehow manage to remove support for 0 in `IntSet`, we could almost implement `find(::BitArray)` as a very fast and lazy conversion to an IntSet. 

I'm not sure how practical that would be since IntSet doesn't implement indexing (and cannot do so efficiently). And it wouldn't satisfy the current meaning of `find` because of that. But if we're clever, it might simplify and improve the performance of logical indexing.

**Comment by quinnj** [comment]
Maybe we just create a new type `IndexSet` that has those behaviors?

**Comment by StefanKarpinski** [comment]
`IndexSet` is really a better name for this type anyway.

**Comment by mbauman** [comment]
Bump!  With inference decoupled from base, this became very easy to debug!  All tests pass now.  The only remaining issue is the performance gap on `in`.

**Comment by dpsanders** [comment]
Bump... I spent time yesterday writing tests for `IntSet`s, only to discover that @mbauman had already done it and completely rewritten everything! +1!  (Maybe you can separate out the tests and get those merged first.)

**Comment by mbauman** [comment]
Shucks.  Sorry for the duplicated effort.  Don't let this long-languishing work in progress hold up your testing work!  Feel free to push your own tests and/or grab tests from here.

If you do grab tests from here, you'll run into some complement IntSet bugs.  I have a hunch that nobody is using complement IntSets very thoroughly.

Maybe we should deprecate the complement functionality?  This is a core piece of functionality that's required for inference (otherwise I don't think it'd be in base).  It could make sense to simplify `IntSet` and move this work into a package for richer `IndexSet`-like functionality.

**Comment by JeffBezanson** [comment]
+1 for removing the complement feature. It was written back in the days when we were bored and just looking for neat stuff to do.

**Comment by StefanKarpinski** [comment]
Agree with removing the complement stuff. I also suspect this type should be renamed to `IndexSet` since it is really not a generic `IntSet` but really a set of indices.

**Comment by dpsanders** [comment]
My effort is at #12247.

The complement functionality is neat, but I can see that it doesn't really belong in Base.
My tests are quite focused around the particular current implementation.

**Comment by ScottPJones** [comment]
I'd agree if this only supports 0 and above and is really only used for indices it should just be added as `IndexSet`, with no complement.  `IntSet` to me implies it needs to allow *any* integer value in the set, not just >= 0.

**Comment by mbauman** [comment]
Perhaps we entirely deprecate `IntSet` and introduce `IndexSet`?  That would allow us to gracefully remove the ability to store zero, too.

**Comment by ScottPJones** [comment]
SGTM :+1:

**Comment by StefanKarpinski** [comment]
I'm entirely convinced that we should remove the ability to store zero...

**Comment by mbauman** [comment]
Wonderful.  Let's do it.  It's actually not as bad as I had been thinking since we already branch for a bounds check.  We can check for zero within that branch with no extra overhead to the other branch.

File: base/intset.jl [hunk 1]
```diff
@@ -1,307 +1,236 @@
...
-    elseif n < 0
-        return false
+    idx = n+1
+    if 1 <= idx <= length(s.bits)
+        unsafe_getindex(s.bits, idx) != s.inverse
     else
-        (s.bits[n>>5 + 1] & (UInt32(1)<<(n&31))) != 0
+        ifelse((idx <= 0) | (idx > typemax(Int)), false, s.inverse)
     end
 end
```
Comment (pabloferz) [inline base/intset.jl]: Aren't you missing the return statements here?

Comment (mbauman) [inline base/intset.jl]: That would probably be more clear, yes, but this works just fine as it is.

The result of the last expression in a function body is used as the return value (if not explicitly specified).  In this case, the last expression is the `if` block.

The result of an `if` block is similarly determined by the last expression of the traversed branch (or `nothing` if no branches were traversed).

Comment (pabloferz) [inline base/intset.jl]: Ok, I knew that the return value of the last expression was used, but I though that an `if` block would have returned `nothing`.

Sorry for the noise.

**Comment by mbauman** [comment]
Closing in favor of https://github.com/JuliaLang/DataStructures.jl/issues/114 (moving IntSet to DataStructures) and https://github.com/JuliaLang/julia/issues/12270 (deprecating complement and stored zeros, on the path to renaming IntSet → IndexSet).

**Pull request closed without merging**